### PR TITLE
add scripts to simulate data

### PR DESCRIPTION
**Pull request by gwaygenomics**
The main function is `getSimulatedExpression()`. It will sample various types of signals including:

1. Groups (based on different normal distributions along given across axes)
2. Nonlinear functions (any transformation to a randomly sampled vector)
3. Cell-types (based on a different proportions of normal distributions across given axes)
4. Presence/absence (random determination if feature is present)
5. Random noise features

I am not trying to comprehensively profile all possible signals present in gene expression data. Just a couple that will help with the evaluations I have in mind.

Some of the evals include:

1. Distinguish groups
2. Identify nonlinear but continuous patterns
3. Robust to noise injection
4. Latent space arithmetic (e.g. `Group 1: Yes Feature` - `Group 1: No Feature` + `Group 2: No Feature` = `Group 2: Yes Feature`?)
5. Generative ability of VAE

I intend to follow up later pull requests comparing Tybalt, ADAGE, PCA, ICA, NMF, and a conditional VAE in these evals.

### CODE REVIEW

**Comment by jaclyn-taroni** [comment]
Do you have any citations for any of these approaches @gwaygenomics? I'm thinking it might be nice to keep track of them here.

**Comment by gwaygenomics** [comment]
> Do you have any citations for any of these approaches @gwaygenomics? I'm thinking it might be nice to keep track of them here.

Do you mean citations for the types of signals, or for the evals?

Many of the simulation studies for the types of signals (that I could easily find) were trying much more complicated things than what I have here. Do you know of any papers that try something similar?

I agree keeping track of these citations is gonna be important.

**Comment by jaclyn-taroni** [comment]
> Do you mean citations for the types of signals, or for the evals?

I meant for the types of signals/simulation approaches. I'm interested in the _rationale_ behind these approaches, which may be beyond the scope of this PR. If there are relevant citations, however, that might be a nice concise way of keeping track of this.

> Do you know of any papers that try something similar?

The [`CellCODE` paper](https://dx.doi.org/10.1093/bioinformatics/btv015) comes to mind for the cell type proportion bit, as does the [`xCell` paper](https://doi.org/10.1186/s13059-017-1349-1).

Also unclear to me (at first glance anyway) if `getSimulatedExpression()` does all of those types of signals at once or not.

**Comment by jaclyn-taroni** [comment]
You might also check out WGCNA -- I think there's functionality to simulate gene expression data with co-expression module structure which you may find useful.

**Comment by gwaygenomics** [comment]
> I meant for the types of signals/simulation approaches. I'm interested in the rationale behind these approaches, which may be beyond the scope of this PR. 

Got it. Yes, this makes sense. I think the rationale is entirely within scope. I started issue #104. I think that is a better place for the discussion than in a PR (which code is primary focus). 

> The CellCODE paper comes to mind for the cell type proportion bit, as does the xCell paper.

Cool! Yeah from my quick glance these are focused on simulating cell-type proportion and use real data (from purified cell lines and assuming an additive model). I also assume an additive model here but don't use real cell types. I think this could be important for the CZI grant, but I am not yet sure it will be important for next steps with Tybalt. If we decide it is, it will belong in a separate PR.

> Also unclear to me (at first glance anyway) if getSimulatedExpression() does all of those types of signals at once or not.

Based on the given input (including how many samples and how many features of each of the proposed 5 signals), the function is able to output all the intended signals at once. They form different features in the n x p output matrix.

> You might also check out WGCNA -- I think there's functionality to simulate gene expression data with co-expression module structure which you may find useful.

I have checked that out before - it looks slightly more complicated than what I have. For our purposes here, I thought it would actually be easier to write my own (that I have more control of). It will be important to cite, however.

File: scripts/util/simulate_expression.R [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+  cell_type_data <- cell_type_params[[1]] + cell_type_params[[2]]
+  return_list <- list(cell_type_data, rand_cell_type_1)
+  return(return_list)
+}
+
+
+getSimulatedExpression <- function(n, mean_df, sd_df, r, func_list, b,
+                                   cell_type_mean_df, cell_type_sd_df,
+                                   zero_one_normalize = TRUE) {
+  # Obtain a matrix with simulated parameters of input size
```
[the remark this answers is not in the document]
Comment (jaclyn-taroni) [reply scripts/util/simulate_expression.R]: Here's what I get about the output:

* Rows are samples, columns are features
* You're going to end up with columns (features) that are associated with the following:
  * Group membership (however many specified by `mean_df` and  `sd_df`)
  * Random noise (number of features specified by `r`)
  * Cell type proportions (2 cell types)
  * Continuous, user-specified (presumably non-linear) functions (`func_list`)
  * "Presence or absence" features (number specified by `b`)
* +/- column-wise normalization
 
And also some "metadata" that will be necessary for evaluation.

File: scripts/util/simulate_expression.R [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  # list of length 2 - 1st element is the group specific feature matrix
+  #                  - 2nd element is a vector of labels ('a', 'b', 'c', etc.)
+
+  group_params <- c()
+  group_name <- letters[1:nrow(mean_matrix)]
+  for (param_idx in 1:ncol(mean_matrix)) {
+    mean_vector <- mean_matrix[, param_idx]
+    sd_vector <- sd_matrix[, param_idx]
+
+    group_vector <- rnorm(n, mean = mean_vector, sd = sd_vector)
```
[the remark this answers is not in the document]
Comment (jaclyn-taroni) [reply scripts/util/simulate_expression.R]: I'm having trouble following why vectors are being passed to `mean` and `sd` here. May become more clear with some additional context.

File: scripts/util/simulate_expression.R [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+  # num_samples - the number of samples to simulate
+  # mean_matrix - a matrix of different group means
+  # sd_matrix - a matrix of different group standard deviations
+  #
+  # The rows of each matrix indicate group specific data (group mean or group
+  # standard deviation) and the columns represent different group features.
+  # (nrow = num groups) (ncol = num features that describe each group)
+  #
+  # Return:
+  # list of length 2 - 1st element is the group specific feature matrix
```
[the remark this answers is not in the document]
Comment (gwaygenomics) [reply scripts/util/simulate_expression.R]: I updated the doc string in the next commit :+1:

File: scripts/util/simulate_expression.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (gwaygenomics) [reply scripts/util/simulate_expression.R]: > Where's n coming from here? When you call it within getSimulatedExpression? I would change to num_samples if that's where this is coming from

good catch. Yeah, that's supposed to be `num_samples`

> I'm having trouble following why vectors are being passed to mean and sd here. May become more clear with some additional context.

`rnorm` can be passed vectors and will sample from them sequentially. For example:

```R
> mean_vector <- c(0, 5)
> sd_vector <- c(1, 1)
> rnorm(10, mean_vector, sd_vector)
 [1]  0.7426884  4.2786223 -1.2607607  5.1591598 -0.8634550  5.1236696 -0.6336891
 [8]  6.0831227  0.3693498  2.7463734
```

So, if I take the columns of the input `mean_matrix` and `sd_matrix`, I can sample the feature for the given input "group". So if my mean matrix is:

| | Feature 1 | Feature 2| 
| :------: | :-----------: | :------------: |
| Group 1 | 0 | 1 |
| Group 2 | 5 | 3 |
| Group 3 | 3 | 6 |

(and assuming an `sd_matrix` of all 1's), the function will return:

```R
> feature_1_mean <- c(0, 5, 3)
> feature_1_sd <- c(1, 1, 1)
> rnorm(10, feature_1_mean, feature_1_sd)
 [1] -0.2012264  5.3283179  3.2768924 -1.1284059  6.4214187  2.0292146  0.6201418
 [8]  5.4884294  3.7513839  0.3292537
```

[the remark this answers is not in the document]
Comment (gwaygenomics) [reply scripts/util/simulate_expression.R]: This note also made me discover a bug in this implementation. In my current usage, I know that it works when `num_samples` is divisible by number of groups. In my example above, there are 3 groups and 10 samples. So I added this fix:

```R
> group_name <- letters[1:length(feature_1_mean)]
> num_samples <- 10
> num_repeat <- num_samples %/% length(group_name)
> num_remainder <- num_samples %% length(group_name)
>  labels <- rep(group_name, num_repeat)
>   
>   if (num_remainder > 0) {
+     labels <- c(labels, group_name[1:num_remainder])
+   }
> labels
 [1] "a" "b" "c" "a" "b" "c" "a" "b" "c" "a"
```

[the remark this answers is not in the document]
Comment (jaclyn-taroni) [reply scripts/util/simulate_expression.R]: Okay, cool I think may have contributed to my confusion! Thanks for the explanation 👍

File: scripts/util/simulate_expression.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (gwaygenomics) [reply scripts/util/simulate_expression.R]: > Needs more info, difficult for me to figure out how I might want to use this with Tybalt

It doesn't necessarily need to be used with Tybalt. But I added in the docstring how one may use the function in this project.

> Here's what I get about the output:

Almost exactly correct! The one thing that could use some clarification (maybe it is understood, but I will be explicit) is the `Cell type proportions (2 cell types)`.

Specifically, there can be more than 2 columns, but only exactly 2 cell-types. For example, if my cell_type_mean matrix had 3 columns:

| | Feature 1 | Feature 2| Feature 3|
| :------: | :-----------: | :------------: |
| Cell-type 1 | 0 | 1 | 0.5 |
| Cell-type 2 | 5 | 3 | 2 |

Then there would be three features (columns) describing cell types. The operation is:

```R
# Randomly (uniformly) sample a number between 0 and 1
cell_type_1 <- runif(1, min=0, max=1)
cell_type_1
[1] 0.8449335

# Subtract to get cell_type_2 proportion
cell_type_2 <- 1 - cell_type_1
[1] 0.1550665

# so assuming a `cell_type_sd` matrix of all zeros, the single sample will be:
c(0, 1, 0.5) *  0.8449335 + c(5, 3, 2) * 0.1550665
[1] 0.7753325 1.3101330 0.7325997
```
I tried to make this more clear in the most recent commit. Thanks!

**Comment by gwaygenomics** [comment]
Thanks for comments @jaclyn-taroni 

> I think I need a little more information/documentation before I can critically evaluate this @gwaygenomics -- want to make sure I'm following

I bolstered the documentation. Does it make more sense now?

**Pull request merged**